test/integration/test_server.c

```diff
@@ -0,0 +1,262 @@
+
+struct fixture
+{
+	char *dirs[N_SERVERS];
+	dqlite_server *servers[N_SERVERS];
+};
+
+static void *setup(const MunitParameter params[], void *user_data)
+{
+	(void)params;
+	(void)user_data;
+	struct fixture *f = munit_malloc(sizeof *f);
+	unsigned i;
+	int rv;
+
+	for (i = 0; i < N_SERVERS; i += 1) {
+		f->dirs[i] = test_dir_setup();
+		rv = dqlite_server_create(f->dirs[i], &f->servers[i]);
+		munit_assert_int(rv, ==, 0);
+	}
+
+	return f;
+}
+
+static void teardown(void *data)
+{
+	struct fixture *f = data;
+	unsigned i;
+
+	for (i = 0; i < N_SERVERS; i += 1) {
+		dqlite_server_destroy(f->servers[i]);
+		test_dir_tear_down(f->dirs[i]);
+	}
+	free(f);
+}
+
+#define PREPARE_FILE(i, name, s, ...)                           \
+	do {                                                    \
+		char path[100];                                 \
+		snprintf(path, 100, "%s/%s", f->dirs[i], name); \
+		FILE *fp = fopen(path, "w+");                   \
+		fprintf(fp, s, __VA_ARGS__);                    \
+		fclose(fp);                                     \
+	} while (0)
+
+#define NODE(x) x
+
+#define NODE0_ID "3297041220608546238"
+
+void start_each_server(struct fixture *f)
+{
+	int rv;
+
+	rv = dqlite_server_set_address(f->servers[0], "127.0.0.1:8880");
+	munit_assert_int(rv, ==, 0);
+	rv = dqlite_server_set_bootstrap(f->servers[0]);
+	munit_assert_int(rv, ==, 0);
+	rv = dqlite_server_set_refresh_period(f->servers[0], 1);
+	munit_assert_int(rv, ==, 0);
+	rv = dqlite_server_start(f->servers[0]);
+	munit_assert_int(rv, ==, 0);
+
+	rv = dqlite_server_set_address(f->servers[1], "127.0.0.1:8881");
+	munit_assert_int(rv, ==, 0);
+	rv = dqlite_server_set_peer_address(f->servers[1], "127.0.0.1:8880");
+	munit_assert_int(rv, ==, 0);
+	rv = dqlite_server_set_refresh_period(f->servers[0], 1);
+	munit_assert_int(rv, ==, 0);
+	rv = dqlite_server_start(f->servers[1]);
+	munit_assert_int(rv, ==, 0);
+
+	rv = dqlite_server_set_address(f->servers[2], "127.0.0.1:8882");
+	munit_assert_int(rv, ==, 0);
+	rv = dqlite_server_set_peer_address(f->servers[2], "127.0.0.1:8880");
+	munit_assert_int(rv, ==, 0);
+	rv = dqlite_server_set_peer_address(f->servers[2], "127.0.0.1:8881");
+	munit_assert_int(rv, ==, 0);
+	rv = dqlite_server_set_refresh_period(f->servers[0], 1);
+	munit_assert_int(rv, ==, 0);
+	rv = dqlite_server_start(f->servers[2]);
+	munit_assert_int(rv, ==, 0);
+}
+
+void stop_each_server(struct fixture *f)
+{
+	int rv;
+
+	rv = dqlite_server_stop(f->servers[2]);
+	munit_assert_int(rv, ==, 0);
+	rv = dqlite_server_stop(f->servers[1]);
+	munit_assert_int(rv, ==, 0);
+	rv = dqlite_server_stop(f->servers[0]);
+	munit_assert_int(rv, ==, 0);
+}
+
+TEST(server, restart_follower, setup, teardown, 0, NULL)
+{
+	struct fixture *f = data;
+	int rv;
+
+	start_each_server(f);
+
+	sleep(2);
+
+	rv = dqlite_server_stop(f->servers[1]);
+	munit_assert_int(rv, ==, 0);
+
+	sleep(2);
+
+	rv = dqlite_server_start(f->servers[1]);
+	munit_assert_int(rv, ==, 0);
+
+	sleep(2);
```

Comment:
Sleeping for 2 seconds seems a relatively long time for a test, also considering that there are several `sleep(2)` calls in the test, and several tests that do the same, adding to the overall execution time of the test suite.

What is the purpose of sleeping in the first place? Exercise the refresh logic?

Ideally the test should monitor for some specific event (e.g. the node store file being refreshed or something like that) and then unblock when that happens, or timeout after a while.

Alternatively, I think lowering the time to something like 100 or 200 milliseconds should be enough to cover whatever this is supposed to test.

Depending on the goal of the test, some other knobs might need to be tweaked, like lowering the refresh period to 50 milliseconds or lowering the network latency with `dqlite_node_set_network_latency_ms()`.

I'm also fine leaving the `sleep(2)` for now if it's useful and the alternatives are complicated, we can improve this later, but wanted to check what the purpose is. and why the specific value of 2 seconds was chosen. Adding a comment would help too.

Reply:
> > What is the purpose of sleeping in the first place? Exercise the refresh logic?
> 
> Yes, that was the idea. I can shorten the refresh period and the sleep, I think waiting explicitly for the refresh logic to run is tricky without introducing some sort of test-only API to invoke a callback whenever refresh occurs (but maybe that would be a good idea anyway?).

As a general rule, for integration tests it'd be great to keep things mostly as a black-box (no test-only APIs, unless we have too).

For this behavior specifically we could for example watch the file system with `inotify` and check that the node store was refreshed correctly.

Or perhaps later down the road when the client will be in place, it will be easier to put in place a test that depends on the refresh logic to succeed (for instance changing the cluster membership so that the client won't be able to find the leader until the node store is refreshed).

I don't think we don't need to go through the `inotify` complication right now, we can revisit this later on if it will become more practical. Reducing the test time by making the refresh period lower is good enough.